services/requester/cross-spork_client.go

```diff
@@ -0,0 +1,155 @@
+	logger zerolog.Logger
+	// this map holds the last heights and clients for each spork
+	sporkClients map[uint64]access.Client
+	access.Client
+}
+
+// NewCrossSporkClient creates a new instance of the multi-spork client. It requires
+// the current spork client and a slice of past spork clients.
+func NewCrossSporkClient(
+	currentSpork access.Client,
+	pastSporks []access.Client,
+	logger zerolog.Logger,
+) (*CrossSporkClient, error) {
+	client := &CrossSporkClient{
+		logger,
+		make(map[uint64]access.Client),
+		currentSpork,
+	}
+
+	for _, sporkClient := range pastSporks {
+		if err := client.addSpork(sporkClient); err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+// addSpork will add a new spork host defined by the last height boundary in that spork.
+func (c *CrossSporkClient) addSpork(client access.Client) error {
+	header, err := client.GetLatestBlockHeader(context.Background(), true)
+	if err != nil {
+		return fmt.Errorf("could not get latest height using the spork client: %w", err)
+	}
+
+	lastHeight := header.Height
+
+	if _, ok := c.sporkClients[lastHeight]; ok {
+		return fmt.Errorf("provided last height already exists")
+	}
+
+	c.sporkClients[lastHeight] = client
+
+	c.logger.Info().
+		Uint64("spork-boundary", lastHeight).
+		Msg("added spork specific client")
+
+	return nil
+}
+
+// IsPastSpork will check if the provided height is contained in the previous sporks.
+func (c *CrossSporkClient) IsPastSpork(height uint64) bool {
+	return len(c.getSporkBoundariesDesc()) > 0 && height <= c.getSporkBoundariesDesc()[0]
+}
+
+// getClientForHeight returns the client for the given height. It starts by using the current spork client,
+// then iteratively checks the upper height boundaries in descending order and returns the last client
+// that still contains the given height within its upper height limit. If no client is found, it returns
+// the current spork client.
+// Please note that even if a client for provided height is found we don't guarantee the data being available
+// because it still might not have access to the height provided, because there might be other sporks with
+// lower height boundaries that we didn't configure for.
+// This would result in the error when using the client to access such data.
+func (c *CrossSporkClient) getClientForHeight(height uint64) access.Client {
+
+	// start by using the current spork client, then iterate all the upper height boundaries
+	// and find the last client that still contains the height in its upper height limit
+	client := c.Client
+	for _, upperBound := range c.getSporkBoundariesDesc() {
+		if upperBound >= height {
+			client = c.sporkClients[upperBound]
+
+			c.logger.Debug().
+				Uint64("spork-boundary", upperBound).
+				Msg("using previous spork client")
+		}
+	}
+
+	return client
+}
+
+// getSporkBoundaries will return descending order of spork height boundaries
+func (c *CrossSporkClient) getSporkBoundariesDesc() []uint64 {
```

Comment:
Minor: We can move this to `NewCrossSporkClient` and cache it, as it is not expected to change after initialization.

Reply:
good point, I moved it into factory.